man-pages/man/man3/getopt.3
'\" t
.\" Copyright, the authors of the Linux man-pages project
.\"
.\" SPDX-License-Identifier: Linux-man-pages-copyleft
.\"
.TH getopt 3 2025-09-21 "Linux man-pages (unreleased)"
.SH NAME
getopt, getopt_long, getopt_long_only,
optarg, optind, opterr, optopt \- Parse command-line options
.SH LIBRARY
Standard C library
.RI ( libc ,\~ \-lc )
.SH SYNOPSIS
.nf
.B #include <unistd.h>
.P
.BI "int getopt(int " argc ", char *" argv [],
.BI "           const char *" optstring );
.P
.BI "extern char *" optarg ;
.BI "extern int " optind ", " opterr ", " optopt ;
.P
.B #include <getopt.h>
.P
.BI "int getopt_long(int " argc ", char *" argv [],
.BI "           const char *" optstring ,
.BI "           const struct option *" longopts ", int *" longindex );
.BI "int getopt_long_only(int " argc ", char *" argv [],
.BI "           const char *" optstring ,
.BI "           const struct option *" longopts ", int *" longindex );
.fi
.P
.RS -4
Feature Test Macro Requirements for glibc (see
.BR feature_test_macros (7)):
.RE
.P
.BR getopt ():
.nf
    _POSIX_C_SOURCE >= 2 || _XOPEN_SOURCE
.fi
.P
.BR getopt_long (),
.BR getopt_long_only ():
.nf
    _GNU_SOURCE
.fi
.SH DESCRIPTION
The
.BR getopt ()
function parses the command-line arguments.
Its arguments
.I argc
and
.I argv
are the argument count and array as passed to the
.IR main ()
function on program invocation.
An element of
.I argv
that starts with \[aq]\-\[aq]
(and is not exactly "\-" or "\-\-")
is an option element.
The characters of this element
(aside from the initial \[aq]\-\[aq]) are option characters.
If
.BR getopt ()
is called repeatedly, it returns successively each of the option characters
from each of the option elements.
.P
The variable
.I optind
is the index of the next element to be processed in
.IR argv .
The system initializes this value to 1.
The caller can reset it to 1 to restart scanning of the same
.IR argv ,
or when scanning a new argument vector.
.P
If
.BR getopt ()
finds another option character, it returns that
character, updating the external variable
.I optind
and a static variable
.I nextchar
so that the next call to
.BR getopt ()
can
resume the scan with the following option character or
.IR argv -element.
.P
If there are no more option characters,
.BR getopt ()
returns \-1.
Then
.I optind
is the index in
.I argv
of the first
.IR argv -element
that is not an option.
.P
.I optstring
is a string containing the legitimate option characters.
A legitimate option character is any visible one byte
.BR ascii (7)
character (for which
.BR isgraph (3)
would return nonzero) that is not \[aq]\-\[aq], \[aq]:\[aq], or \[aq];\[aq].
If such a
character is followed by a colon, the option requires an argument, so
.BR getopt ()
places a pointer to the following text in the same
.IR argv -element,
or the text of the following
.IR argv -element,
in
.IR optarg .
Two colons mean an option takes
an optional arg;
if there is text in the current
.IR argv -element
(i.e., in the same word as the option name itself, for example, "\-oarg"),
then it is returned in
.IR optarg ,
otherwise
.I optarg
is set to zero.
This is a GNU extension.
If
.I optstring
contains
.B W
followed by a semicolon, then
.B \-W foo
is treated as the long option
.BR \-\-foo .
(The
.B \-W
option is reserved by POSIX.2 for implementation extensions.)
This behavior is a GNU extension, not available with libraries before
glibc 2.
.P
By default,
.BR getopt ()
permutes the contents of
.I argv
as it scans,
so that eventually all the nonoptions are at the end.
Two other scanning modes are also implemented.
If the first character of
.I optstring
is \[aq]+\[aq] or the environment variable
.B POSIXLY_CORRECT
is set, then option processing stops as soon as a nonoption argument is
encountered.
If \[aq]+\[aq] is not the first character of
.IR optstring ,
it is treated as a normal option.
If
.B POSIXLY_CORRECT
behaviour is required in this case
.I optstring
will contain two \[aq]+\[aq] symbols.
If the first character of
.I optstring
is \[aq]\-\[aq],
then each nonoption
.IR argv -element
is handled as if it were the argument of
an option with character code 1.
(This is used by programs that were
written to expect options and other
.IR argv -elements
in any order
and that care about the ordering of the two.)
The special argument "\-\-" forces an end of option-scanning regardless
of the scanning mode.
.P
While processing the option list,
.BR getopt ()
can detect two kinds of errors:
(1) an option character that was not specified in
.I optstring
and (2) a missing option argument
(i.e., an option at the end of the command line without an expected argument).
Such errors are handled and reported as follows:
.IP \[bu] 3
By default,
.BR getopt ()
prints an error message on standard error,
places the erroneous option character in
.IR optopt ,
and returns \[aq]?\[aq] as the function result.
.IP \[bu]
If the caller has set the global variable
.I opterr
to zero, then
.BR getopt ()
does not print an error message.
The caller can determine that there was an error by testing whether
the function return value is \[aq]?\[aq].
(By default,
.I opterr
has a nonzero value.)
.IP \[bu]
If the first character
(following any optional \[aq]+\[aq] or \[aq]\-\[aq] described above)
of
.I optstring
is a colon (\[aq]:\[aq]), then
.BR getopt ()
likewise does not print an error message.
In addition, it returns \[aq]:\[aq] instead of \[aq]?\[aq] to
indicate a missing option argument.
This allows the caller to distinguish the two different types of errors.
.\"
.SS getopt_long() and getopt_long_only()
The
.BR getopt_long ()
function works like
.BR getopt ()
except that it also accepts long options, started with two dashes.
(If the program accepts only long options, then
.I optstring
should be specified as an empty string (""), not NULL.)
Long option names may be abbreviated if the abbreviation is
unique or is an exact match for some defined option.
A long option
may take a parameter, of the form
.B \-\-arg=param
or
.BR "\-\-arg param" .
.P
.I longopts
is a pointer to the first element of an array of
.I struct option
declared in
.I <getopt.h>
as
.P
.in +4n
.EX
struct option {
    const char *name;
    int         has_arg;
    int        *flag;
    int         val;
};
.EE
.in
.P
The meanings of the different fields are:
.TP
.I name
is the name of the long option.
.TP
.I has_arg
is:
.B no_argument
(or 0) if the option does not take an argument;
.B required_argument
(or 1) if the option requires an argument;
or
.B optional_argument
(or 2) if the option takes an optional argument.
.TP
.I flag
specifies how results are returned for a long option.
If
.I flag
is NULL, then
.BR getopt_long ()
returns
.IR val .
(For example, the calling program may set
.I val
to the equivalent short
option character.)
Otherwise,
.BR getopt_long ()
returns 0, and
.I flag
points to a variable which is set to
.I val
if the option is found,
but left unchanged if the option is not found.
.TP
.I val
is the value to return, or to load into the variable pointed
to by
.IR flag .
.P
The last element of the array has to be filled with zeros.
.P
If
.I longindex
is not NULL, it
points to a variable which is set to the index of the long option relative to
.IR longopts .
.P
.BR getopt_long_only ()
is like
.BR getopt_long (),
but \[aq]\-\[aq] as well
as "\-\-" can indicate a long option.
If an option that starts with \[aq]\-\[aq]
(not "\-\-") doesn't match a long option, but does match a short option,
it is parsed as a short option instead.
.SH RETURN VALUE
If an option was successfully found, then
.BR getopt ()
returns the option character.
If all command-line options have been parsed, then
.BR getopt ()
returns \-1.
If
.BR getopt ()
encounters an option character that was not in
.IR optstring ,
then \[aq]?\[aq] is returned.
If
.BR getopt ()
encounters an option with a missing argument,
then the return value depends on the first character in
.IR optstring :
if it is \[aq]:\[aq], then \[aq]:\[aq] is returned;
otherwise \[aq]?\[aq] is returned.
.P
.BR getopt_long ()
and
.BR getopt_long_only ()
also return the option
character when a short option is recognized.
For a long option, they
return
.I val
if
.I flag
is NULL, and 0 otherwise.
Error and \-1 returns are the same as for
.BR getopt (),
plus \[aq]?\[aq] for an
ambiguous match or an extraneous parameter.
.SH ENVIRONMENT
.TP
.B POSIXLY_CORRECT
If this is set, then option processing stops as soon as a nonoption
argument is encountered.
.SH ATTRIBUTES
For an explanation of the terms used in this section, see
.BR attributes (7).
.TS
allbox;
lb lb lbx
l l l.
Interface	Attribute	Value
T{
.na
.nh
.BR getopt (),
.BR getopt_long (),
.BR getopt_long_only ()
T}	Thread safety	T{
.na
.nh
MT-Unsafe race:getopt env
T}
.TE
.SH VERSIONS
POSIX specifies that the
.I argv
array argument should be
.IR const ,
but these functions permute its elements
unless the environment variable
.B POSIXLY_CORRECT
is set.
.I const
is used in the actual prototype to be compatible with other systems;
however, this page doesn't show the qualifier,
to avoid confusing readers.
.SH STANDARDS
.TP
.BR getopt ()
POSIX.1-2008.
.TP
.BR getopt_long ()
.TQ
.BR getopt_long_only ()
GNU.
.IP
The use of \[aq]+\[aq] and \[aq]\-\[aq] in
.I optstring
is a GNU extension.
.SH HISTORY
.TP
.BR getopt ()
POSIX.1-2001, and POSIX.2.
.P
On some older implementations,
.BR getopt ()
was declared in
.IR <stdio.h> .
SUSv1 permitted the declaration to appear in either
.I <unistd.h>
or
.IR <stdio.h> .
POSIX.1-1996 marked the use of
.I <stdio.h>
for this purpose as LEGACY.
POSIX.1-2001 does not require the declaration to appear in
.IR <stdio.h> .
.P
Very old versions of glibc were affected by a
.UR https://\:sourceware.org/\:git/\:?p=glibc.git;a=commitdiff;h=bf079e19f50d64aa5e05
.BI _ PID _GNU_nonoption_argv_flags_
environment variable
.UE .
.SH NOTES
A program that scans multiple argument vectors,

man-pages/Changes
==================== Changes in man-pages-6.16 ====================

Released: 2025-10-29, Aldaya


New and rewritten pages
-----------------------

man2/
	fsconfig.2
	fsmount.2
	fsopen.2
	fspick.2
	move_mount.2
	open_tree.2

man2const/
	PR_FUTEX_HASH_GET_SLOTS.2const
	PR_FUTEX_HASH_SET_SLOTS.2const

man2type/
	mount_attr.2type			(split from mount_setattr(2))

man3/
	_Fork.3
	countof.3
	memeq.3
	streq.3


Newly documented interfaces in existing pages
---------------------------------------------

man2/
	mount_setattr.2
		glibc wrapper

	pidfd_open.2
		PIDFD_THREAD

	pidfd_send_signal.2
		PIDFD_SIGNAL_THREAD
		PIDFD_SIGNAL_THREAD_GROUP
		PIDFD_SIGNAL_PROCESS_GROUP

	quotactl.2
		quotactl_fd(2)

	readv.2
		RWF_DONTCACHE

	sigaction.2
		SI_ASYNCNL
		FPE_FLTUNK
		FPE_CONDTRAP
		SEGV_ACCADI
		SEGV_ADIDERR
		SEGV_ADIPERR
		SEGV_MTEAERR
		SEGV_MTESERR
		SEGV_CPERR
		TRAP_UNK
		TRAP_PERF
		SYS_USER_DISPATCH
		EMT_TAGOVF

man3/
	abs.3
		umaxabs(3)

	opendir.3
		ENAMETOOLONG

man7/
	operator.7
		(type){}
		alignof
		_Countof

	rtnetlink.7
		RTAX_UNSPEC
		RTAX_LOCK
		RTAX_MTU
		RTAX_WINDOW
		RTAX_RTT
		RTAX_RTTVAR
		RTAX_SSTHRESH
		RTAX_CWND
		RTAX_ADVMSS
		RTAX_REORDERING
		RTAX_HOPLIMIT
		RTAX_INITCWND
		RTAX_FEATURES
		RTAX_RTO_MIN
		RTAX_INITRWND
		RTAX_QUICKACK
		RTAX_CC_ALGO
		RTAX_FASTOPEN_NO_COOKIE

	sched.7
		CONFIG_PREEMPT
		CONFIG_PREEMPT_LAZY

	tcp.7
		TCP_SAVE_SYN
		TCP_SAVED_SYN


New and changed links
---------------------

man2/
	open_tree_attr.2			(open_tree(2))
	quotactl_fd.2				(quotactl(2))
man3/
	umaxabs.3				(abs(3))


Global changes
--------------

-  man/
   -  Update documentation for POSIX.1-2024 (still more work to do).
   -  SYNOPSIS: Use GNU forward-declarations of parameters for sizes of
      array parameters.
   -  Improve style of source code for tbl(1) tables.
   -  man2/: Use the common name 'fd' instead of 'fildes'.
   -  Use semantic newlines.
   -  EXAMPLES
      -  Use err(3) and errc(3bsd) instead of similar macros.
      -  Fix includes (thanks to iwyu(1)).
      -  Use NITEMS() consistently.
      -  Don't name unused parameters.
-  share/mk/
   -  Add support for the chapter man3attr.
   -  help: Document the usual targets.
   -  lint-man-blank: Add target to lint about blank lines.
   -  lint-man-ws: Diagnose spurious use of white space.
   -  lint-man-quote: Diagnose an unmatched quote.
   -  lint-man-poems: Diagnose (lack of) semantic newlines.
   -  lint-man-dash: Diagnose unescaped dashes
   -  Merge handling of man(7) and mdoc(7) pages.
   -  lint, build, check: Move exceptions to separate files.
   -  Remove unused variables.
   -  Unify local variable names.
   -  Silence diagnostics about unnamed parameters.
   -  configure/build-depends/:
      -  Add awk/awk.
      -  Add pcre2-utils/pcre2grep.


Changes to individual files
---------------------------

-  src/bin/
   -  sortman(1): Fix order of intro(*) pages for subsections.
   -  diffman-git(1): Show sections as hunk context.

The manual pages and other files in the repository have been improved
beyond what this changelog covers.  To learn more about changes applied
to individual pages, or the authors of changes, use git(1).


src/wget2/mandb/lib/getopt.c
/* Getopt for GNU.
   NOTE: getopt is now part of the C library, so if you don't know what
   "Keep this file name-space clean" means, talk to roland@gnu.ai.mit.edu
   before changing it!

   Copyright (C) 1987, 88, 89, 90, 91, 92, 93, 94
   	Free Software Foundation, Inc.

   This program is free software; you can redistribute it and/or modify it
   under the terms of the GNU General Public License as published by the
   Free Software Foundation; either version 2, or (at your option) any
   later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.  */

/* This tells Alpha OSF/1 not to define a getopt prototype in <stdio.h>.
   Ditto for AIX 3.2 and <stdlib.h>.  */
#ifndef _NO_PROTO
#define _NO_PROTO
#endif

#ifdef HAVE_CONFIG_H
#  include "config.h"
#endif

#if !defined (__STDC__) || !__STDC__
/* This is a separate conditional since some stdc systems
   reject `defined (const)'.  */
#ifndef const
#define const
#endif
#endif

#include <stdio.h>

/* Comment out all this code if we are using the GNU C Library, and are not
   actually compiling the library itself.  This code is part of the GNU C
   Library, but also included in many other GNU distributions.  Compiling
   and linking in this code is a waste when using the GNU C library
   (especially if it is a shared library).  Rather than having every GNU
   program understand `configure --with-gnu-libc' and omit the object files,
   it is simpler to just do this in the source for each such file.  */

#if defined (_LIBC) || !defined (__GNU_LIBRARY__)


/* This needs to come after some library #include
   to get __GNU_LIBRARY__ defined.  */
#ifdef	__GNU_LIBRARY__
/* Don't include stdlib.h for non-GNU C libraries because some of them
   contain conflicting prototypes for getopt.  */
#include <stdlib.h>
#endif	/* GNU C library.  */

/* This version of `getopt' appears to the caller like standard Unix `getopt'
   but it behaves differently for the user, since it allows the user
   to intersperse the options with the other arguments.

   As `getopt' works, it permutes the elements of ARGV so that,
   when it is done, all the options precede everything else.  Thus
   all application programs are extended to handle flexible argument order.

   Setting the environment variable POSIXLY_CORRECT disables permutation.
   Then the behavior is completely standard.

   GNU application programs can use a third alternative mode in which
   they can distinguish the relative order of options and other arguments.  */

#include "getopt.h"

/* For communication from `getopt' to the caller.
   When `getopt' finds an option that takes an argument,
   the argument value is returned here.
   Also, when `ordering' is RETURN_IN_ORDER,
   each non-option ARGV-element is returned here.  */

char *optarg = NULL;

/* Index in ARGV of the next element to be scanned.
   This is used for communication to and from the caller
   and for communication between successive calls to `getopt'.

   On entry to `getopt', zero means this is the first call; initialize.

   When `getopt' returns EOF, this is the index of the first of the
   non-option elements that the caller should itself scan.

   Otherwise, `optind' communicates from one call to the next
   how much of ARGV has been scanned so far.  */

/* XXX 1003.2 says this must be 1 before any call.  */
int optind = 0;

/* The next char to be scanned in the option-element
   in which the last option character we returned was found.
   This allows us to pick up the scan where we left off.

   If this is zero, or a null string, it means resume the scan
   by advancing to the next ARGV-element.  */

static char *nextchar;

/* Callers store zero here to inhibit the error message
   for unrecognized options.  */

int opterr = 1;

/* Set to an option character which was unrecognized.
   This must be initialized on some systems to avoid linking in the
   system's own getopt implementation.  */

int optopt = '?';

/* Describe how to deal with options that follow non-option ARGV-elements.

   If the caller did not specify anything,
   the default is REQUIRE_ORDER if the environment variable
   POSIXLY_CORRECT is defined, PERMUTE otherwise.

   REQUIRE_ORDER means don't recognize them as options;
   stop option processing when the first non-option is seen.
   This is what Unix does.
   This mode of operation is selected by either setting the environment
   variable POSIXLY_CORRECT, or using `+' as the first character
   of the list of option characters.

   PERMUTE is the default.  We permute the contents of ARGV as we scan,
   so that eventually all the non-options are at the end.  This allows options
   to be given in any order, even with programs that were not written to
   expect this.

   RETURN_IN_ORDER is an option available to programs that were written
   to expect options and other ARGV-elements in any order and that care about
   the ordering of the two.  We describe each non-option ARGV-element
   as if it were the argument of an option with character code 1.
   Using `-' as the first character of the list of option characters
   selects this mode of operation.

   The special argument `--' forces an end of option-scanning regardless
   of the value of `ordering'.  In the case of RETURN_IN_ORDER, only
   `--' can cause `getopt' to return EOF with `optind' != ARGC.  */

static enum
{
  REQUIRE_ORDER, PERMUTE, RETURN_IN_ORDER
} ordering;

/* Value of POSIXLY_CORRECT environment variable.  */
static char *posixly_correct;

#ifdef	__GNU_LIBRARY__
/* We want to avoid inclusion of string.h with non-GNU libraries
   because there are many ways it can cause trouble.
   On some systems, it contains special magic macros that don't work
   in GCC.  */
#include <string.h>
#define	my_index	strchr
#else

/* Avoid depending on library functions or files
   whose names are inconsistent.  */

char *getenv ();

static char *
my_index (str, chr)
     const char *str;
     int chr;
{
  while (*str)
    {
      if (*str == chr)
	return (char *) str;
      str++;
    }
  return 0;
}

/* If using GCC, we can safely declare strlen this way.
   If not using GCC, it is ok not to declare it.  */
#ifdef __GNUC__
/* Note that Motorola Delta 68k R3V7 comes with GCC but not stddef.h.
   That was relevant to code that was here before.  */
#if !defined (__STDC__) || !__STDC__
/* gcc with -traditional declares the built-in strlen to return int,
   and has done so at least since version 2.4.5. -- rms.  */
extern int strlen (const char *);
#endif /* not __STDC__ */
#endif /* __GNUC__ */

#endif /* not __GNU_LIBRARY__ */

/* Handle permutation of arguments.  */

/* Describe the part of ARGV that contains non-options that have
   been skipped.  `first_nonopt' is the index in ARGV of the first of them;
   `last_nonopt' is the index after the last of them.  */

static int first_nonopt;
static int last_nonopt;

/* Exchange two adjacent subsequences of ARGV.
   One subsequence is elements [first_nonopt,last_nonopt)
   which contains all the non-options that have been skipped so far.
   The other is elements [last_nonopt,optind), which contains all
   the options processed since those non-options were skipped.

   `first_nonopt' and `last_nonopt' are relocated so that they describe
   the new indices of the non-options in ARGV after they are moved.  */

static void
exchange (argv)
     char **argv;
{
  int bottom = first_nonopt;
  int middle = last_nonopt;
  int top = optind;
  char *tem;

  /* Exchange the shorter segment with the far end of the longer segment.
     That puts the shorter segment into the right place.
     It leaves the longer segment in the right place overall,
     but it consists of two parts that need to be swapped next.  */

  while (top > middle && middle > bottom)
    {
      if (top - middle > middle - bottom)
	{
	  /* Bottom segment is the short one.  */
	  int len = middle - bottom;
	  register int i;

	  /* Swap it with the top part of the top segment.  */
	  for (i = 0; i < len; i++)
	    {
	      tem = argv[bottom + i];
	      argv[bottom + i] = argv[top - (middle - bottom) + i];
	      argv[top - (middle - bottom) + i] = tem;
	    }
	  /* Exclude the moved bottom segment from further swapping.  */
	  top -= len;
	}
      else
	{
	  /* Top segment is the short one.  */
	  int len = top - middle;
	  register int i;

	  /* Swap it with the bottom part of the bottom segment.  */
	  for (i = 0; i < len; i++)
	    {
	      tem = argv[bottom + i];
	      argv[bottom + i] = argv[middle + i];
	      argv[middle + i] = tem;
	    }
	  /* Exclude the moved top segment from further swapping.  */
	  bottom += len;
	}
    }

  /* Update records for the slots the non-options now occupy.  */

  first_nonopt += (optind - last_nonopt);
  last_nonopt = optind;
}

/* Initialize the internal data when the first call is made.  */

static const char *
_getopt_initialize (optstring)
     const char *optstring;
{
  /* Start processing options with ARGV-element 1 (since ARGV-element 0
     is the program name); the sequence of previously skipped
     non-option ARGV-elements is empty.  */

  first_nonopt = last_nonopt = optind = 1;

  nextchar = NULL;

  posixly_correct = getenv ("POSIXLY_CORRECT");

  /* Determine how to handle the ordering of options and nonoptions.  */

  if (optstring[0] == '-')
    {
      ordering = RETURN_IN_ORDER;
      ++optstring;
    }
  else if (optstring[0] == '+')
    {
      ordering = REQUIRE_ORDER;
      ++optstring;
    }
  else if (posixly_correct != NULL)
    ordering = REQUIRE_ORDER;
  else
    ordering = PERMUTE;

  return optstring;
}

/* Scan elements of ARGV (whose length is ARGC) for option characters
   given in OPTSTRING.

   If an element of ARGV starts with '-', and is not exactly "-" or "--",
   then it is an option element.  The characters of this element
   (aside from the initial '-') are option characters.  If `getopt'
   is called repeatedly, it returns successively each of the option characters
   from each of the option elements.

   If `getopt' finds another option character, it returns that character,
   updating `optind' and `nextchar' so that the next call to `getopt' can
   resume the scan with the following option character or ARGV-element.

   If there are no more option characters, `getopt' returns `EOF'.
   Then `optind' is the index in ARGV of the first ARGV-element
   that is not an option.  (The ARGV-elements have been permuted
   so that those that are not options now come last.)

   OPTSTRING is a string containing the legitimate option characters.
   If an option character is seen that is not listed in OPTSTRING,
   return '?' after printing an error message.  If you set `opterr' to
   zero, the error message is suppressed but we still return '?'.

   If a char in OPTSTRING is followed by a colon, that means it wants an arg,
   so the following text in the same ARGV-element, or the text of the following
   ARGV-element, is returned in `optarg'.  Two colons mean an option that
   wants an optional arg; if there is text in the current ARGV-element,
   it is returned in `optarg', otherwise `optarg' is set to zero.

   If OPTSTRING starts with `-' or `+', it requests different methods of
   handling the non-option ARGV-elements.
   See the comments about RETURN_IN_ORDER and REQUIRE_ORDER, above.

   Long-named options begin with `--' instead of `-'.
   Their names may be abbreviated as long as the abbreviation is unique
   or is an exact match for some defined option.  If they have an
   argument, it follows the option name in the same ARGV-element, separated
   from the option name by a `=', or else the in next ARGV-element.
   When `getopt' finds a long-named option, it returns 0 if that option's
   `flag' field is nonzero, the value of the option's `val' field
   if the `flag' field is zero.

   The elements of ARGV aren't really const, because we permute them.
   But we pretend they're const in the prototype to be compatible
   with other systems.

   LONGOPTS is a vector of `struct option' terminated by an
   element containing a name which is zero.

   LONGIND returns the index in LONGOPT of the long-named option found.
   It is only valid when a long-named option has been found by the most
   recent call.

   If LONG_ONLY is nonzero, '-' as well as '--' can introduce
   long-named options.  */

int
_getopt_internal (argc, argv, optstring, longopts, longind, long_only)
     int argc;
     char *const *argv;
     const char *optstring;
     const struct option *longopts;
     int *longind;
     int long_only;
{
  optarg = NULL;

  if (optind == 0)
    optstring = _getopt_initialize (optstring);

  if (nextchar == NULL || *nextchar == '\0')
    {
      /* Advance to the next ARGV-element.  */

      if (ordering == PERMUTE)
	{
	  /* If we have just processed some options following some non-options,
	     exchange them so that the options come first.  */

	  if (first_nonopt != last_nonopt && last_nonopt != optind)
	    exchange ((char **) argv);
	  else if (last_nonopt != optind)
	    first_nonopt = optind;

	  /* Skip any additional non-options
	     and extend the range of non-options previously skipped.  */

	  while (optind < argc
		 && (argv[optind][0] != '-' || argv[optind][1] == '\0'))
	    optind++;
	  last_nonopt = optind;
	}

      /* The special ARGV-element `--' means premature end of options.
	 Skip it like a null option,
	 then exchange with previous non-options as if it were an option,
	 then skip everything else like a non-option.  */

      if (optind != argc && !strcmp (argv[optind], "--"))
	{
	  optind++;

	  if (first_nonopt != last_nonopt && last_nonopt != optind)
	    exchange ((char **) argv);
	  else if (first_nonopt == last_nonopt)
	    first_nonopt = optind;
	  last_nonopt = argc;

	  optind = argc;
	}

      /* If we have done all the ARGV-elements, stop the scan
	 and back over any non-options that we skipped and permuted.  */

      if (optind == argc)
	{
	  /* Set the next-arg-index to point at the non-options
	     that we previously skipped, so the caller will digest them.  */
	  if (first_nonopt != last_nonopt)
	    optind = first_nonopt;
	  return EOF;
	}

      /* If we have come to a non-option and did not permute it,
	 either stop the scan or describe it to the caller and pass it by.  */

      if ((argv[optind][0] != '-' || argv[optind][1] == '\0'))
	{
	  if (ordering == REQUIRE_ORDER)
	    return EOF;
	  optarg = argv[optind++];
	  return 1;
	}

      /* We have found another option-ARGV-element.
	 Skip the initial punctuation.  */

      nextchar = (argv[optind] + 1
		  + (longopts != NULL && argv[optind][1] == '-'));
    }

  /* Decode the current option-ARGV-element.  */

  /* Check whether the ARGV-element is a long option.

     If long_only and the ARGV-element has the form "-f", where f is
     a valid short option, don't consider it an abbreviated form of
     a long option that starts with f.  Otherwise there would be no
     way to give the -f short option.

     On the other hand, if there's a long option "fubar" and
     the ARGV-element is "-fu", do consider that an abbreviation of
     the long option, just like "--fu", and not "-f" with arg "u".

     This distinction seems to be the most useful approach.  */

  if (longopts != NULL
      && (argv[optind][1] == '-'
	  || (long_only && (argv[optind][2] || !my_index (optstring, argv[optind][1])))))
    {
      char *nameend;
      const struct option *p;
      const struct option *pfound = NULL;
      int exact = 0;
      int ambig = 0;
      int indfound;
      int option_index;

      for (nameend = nextchar; *nameend && *nameend != '='; nameend++)
	/* Do nothing.  */ ;

      /* Test all long options for either exact match
	 or abbreviated matches.  */
      for (p = longopts, option_index = 0; p->name; p++, option_index++)
	if (!strncmp (p->name, nextchar, nameend - nextchar))
	  {
	    if (nameend - nextchar == strlen (p->name))
	      {
		/* Exact match found.  */
		pfound = p;
		indfound = option_index;
		exact = 1;
		break;
	      }
	    else if (pfound == NULL)
	      {
		/* First nonexact match found.  */
		pfound = p;
		indfound = option_index;
	      }
	    else
	      /* Second or later nonexact match found.  */
	      ambig = 1;
	  }

      if (ambig && !exact)
	{
	  if (opterr)
	    fprintf (stderr, "%s: option `%s' is ambiguous\n",
		     argv[0], argv[optind]);
	  nextchar += strlen (nextchar);
	  optind++;
	  return '?';
	}

      if (pfound != NULL)
	{
	  option_index = indfound;
	  optind++;
	  if (*nameend)
	    {
	      /* Don't test has_arg with >, because some C compilers don't
		 allow it to be used on enums.  */
	      if (pfound->has_arg)
		optarg = nameend + 1;
	      else
		{
		  if (opterr)
		    {
		      if (argv[optind - 1][1] == '-')
			/* --option */
			fprintf (stderr,
				 "%s: option `--%s' doesn't allow an argument\n",
				 argv[0], pfound->name);
		      else
			/* +option or -option */
			fprintf (stderr,
			     "%s: option `%c%s' doesn't allow an argument\n",
			     argv[0], argv[optind - 1][0], pfound->name);
		    }
		  nextchar += strlen (nextchar);
		  return '?';
		}
	    }
	  else if (pfound->has_arg == 1)
	    {
	      if (optind < argc)
		optarg = argv[optind++];
	      else
		{
		  if (opterr)
		    fprintf (stderr, "%s: option `%s' requires an argument\n",
			     argv[0], argv[optind - 1]);
		  nextchar += strlen (nextchar);
		  return optstring[0] == ':' ? ':' : '?';
		}
	    }
	  nextchar += strlen (nextchar);
	  if (longind != NULL)
	    *longind = option_index;
	  if (pfound->flag)
	    {
	      *(pfound->flag) = pfound->val;
	      return 0;
	    }
	  return pfound->val;
	}

      /* Can't find it as a long option.  If this is not getopt_long_only,
	 or the option starts with '--' or is not a valid short
	 option, then it's an error.
	 Otherwise interpret it as a short option.  */
      if (!long_only || argv[optind][1] == '-'
	  || my_index (optstring, *nextchar) == NULL)
	{
	  if (opterr)
	    {
	      if (argv[optind][1] == '-')
		/* --option */
		fprintf (stderr, "%s: unrecognized option `--%s'\n",
			 argv[0], nextchar);
	      else
		/* +option or -option */
		fprintf (stderr, "%s: unrecognized option `%c%s'\n",
			 argv[0], argv[optind][0], nextchar);
	    }
	  nextchar = (char *) "";
	  optind++;
	  return '?';
	}
    }

  /* Look at and handle the next short option-character.  */

  {
    char c = *nextchar++;
    char *temp = my_index (optstring, c);

    /* Increment `optind' when we start to process its last character.  */
    if (*nextchar == '\0')
      ++optind;

    if (temp == NULL || c == ':')
      {
	if (opterr)
	  {
	    if (posixly_correct)
	      /* 1003.2 specifies the format of this message.  */
	      fprintf (stderr, "%s: illegal option -- %c\n", argv[0], c);
	    else
	      fprintf (stderr, "%s: invalid option -- %c\n", argv[0], c);
	  }
	optopt = c;
	return '?';
      }
    if (temp[1] == ':')
      {
	if (temp[2] == ':')
	  {
	    /* This is an option that accepts an argument optionally.  */
	    if (*nextchar != '\0')
	      {
		optarg = nextchar;
		optind++;
	      }
	    else
	      optarg = NULL;
	    nextchar = NULL;
	  }
	else
	  {
	    /* This is an option that requires an argument.  */
	    if (*nextchar != '\0')
	      {
		optarg = nextchar;
		/* If we end this ARGV-element by taking the rest as an arg,
		   we must advance to the next element now.  */
		optind++;
	      }
	    else if (optind == argc)
	      {
		if (opterr)
		  {
		    /* 1003.2 specifies the format of this message.  */
		    fprintf (stderr, "%s: option requires an argument -- %c\n",
			     argv[0], c);
		  }
		optopt = c;
		if (optstring[0] == ':')
		  c = ':';
		else
		  c = '?';
	      }
	    else
	      /* We already incremented `optind' once;
		 increment it again when taking next ARGV-elt as argument.  */
	      optarg = argv[optind++];
	    nextchar = NULL;
	  }
      }
    return c;
  }
}

int
getopt (argc, argv, optstring)
     int argc;
     char *const *argv;
     const char *optstring;
{
  return _getopt_internal (argc, argv, optstring,
			   (const struct option *) 0,
			   (int *) 0,
			   0);
}

#endif	/* _LIBC or not __GNU_LIBRARY__.  */

#ifdef TEST

/* Compile with -DTEST to make an executable for use in testing
   the above definition of `getopt'.  */

int
main (argc, argv)
     int argc;
     char **argv;
{
  int c;
  int digit_optind = 0;

  while (1)
    {
      int this_option_optind = optind ? optind : 1;

      c = getopt (argc, argv, "abc:d:0123456789");
      if (c == EOF)
	break;

      switch (c)
	{
	case '0':
	case '1':
	case '2':
	case '3':
	case '4':
	case '5':
	case '6':
	case '7':
	case '8':
	case '9':
	  if (digit_optind != 0 && digit_optind != this_option_optind)
	    printf ("digits occur in two different argv-elements.\n");
	  digit_optind = this_option_optind;
	  printf ("option %c\n", c);
	  break;

	case 'a':
	  printf ("option a\n");
	  break;

	case 'b':
	  printf ("option b\n");
	  break;

	case 'c':
	  printf ("option c with value `%s'\n", optarg);
	  break;

	case '?':
	  break;

	default:
	  printf ("?? getopt returned character code 0%o ??\n", c);
	}
    }

  if (optind < argc)
    {
      printf ("non-option ARGV-elements: ");
      while (optind < argc)
	printf ("%s ", argv[optind++]);
      printf ("\n");
    }

  exit (0);
}

#endif /* TEST */
